### DISCUSSION TITLE
Problem with headless session

### DISCUSSION THREAD

**Discussion by roland-5** [any1/wayvnc#84] [Q&A]
Hi,

I wanted make headless session over ssh with my computer at work. My wayvnc config:
```
address=0.0.0.0
enable_auth=true
username=luser
password=p455w0rd
private_key_file=~/.config/wayvnc/key.pem
certificate_file=~/.config/wayvnc/cert.pem
```

When I type wayvnc, there is this error:
```
error: XDG_RUNTIME_DIR not set in the environment.
ERROR: Failed to initialise wayland
```

My .pam_environment
```
SSH_AUTH_SOCK DEFAULT="${XDG_RUNTIME_DIR}/ssh-agent.socket"
WLR_BACKENDS=headless
WLR_LIBINPUT_NO_DEVICES=1
XDG_RUNTIME_DIR=/run/user/$(id -u)
```

I put this WLR's in wrong file?

**Comment by Flakebi** [comment]
Are you sure `XDG_RUNTIME_DIR` is set in your environment? Does `echo $XDG_RUNTIME_DIR` print something?

I had the same problem when starting wayvnc from inside tmux, which was running as a systemd service and didn’t inherit my user environment. Setting `XDG_RUNTIME_DIR` manually to `/run/user/1000` worked fine.

**Reply by roland-5** [reply]
Huh, you were right, it's problem with environment. When I type
```
syntezoid@workcomputer ~ 22:25 ❯ echo $XDG_RUNTIME_DIR

```
There is just blank space. I changed XDG_RUNTIME_DIR to XDG_RUNTIME_DIR=/run/user/1000 but effect is the same - blank space, after echoing. I rebooted computer, but nothing changed. It's strange, because I checked this on my home computer and it show me
```
syntezoid@homesynt ~ 27m52s/22:55 ❯ echo $XDG_RUNTIME_DIR
/run/user/1000
```

I installed system on both computers at the same day.

**Reply by Flakebi** [reply]
What happens if you do `export XDG_RUNTIME_DIR=/run/user/1000` and then run wayvnc or print the variable? If that doesn’t work, does `env XDG_RUNTIME_DIR=/run/user/1000 wayvnc` work?

I never used `.pam_environment`, but it looks like it doesn’t support bash syntax<sup>[1](https://linux.die.net/man/5/pam_env.conf)</sup>, so it should probably look like this:
```bash
SSH_AUTH_SOCK DEFAULT="${XDG_RUNTIME_DIR}/ssh-agent.socket"
WLR_BACKENDS DEFAULT=headless
WLR_LIBINPUT_NO_DEVICES DEFAULT=1
XDG_RUNTIME_DIR DEFAULT=/run/user/1000
```

Or maybe you don’t even need the `XDG_RUNTIME_DIR` in there, it gets set automatically for me, without having that file.

**Reply by roland-5** [reply]
I should mention it, I use fish as default shell. Maybe putting WLR commands to pam_enviroment is bad idea, but I'm not sure where else to put these two lines. At home computer, I don't even have .pam_environment file, and like you said, it gets me automatically to see XDG_RUNTIME_DIR.

```
syntezoid@workcomputer ~ 21:32 ❯ export XDG_RUNTIME_DIR=/run/user/1000
syntezoid@workcomputer ~ 21:32 ❯ echo $XDG_RUNTIME_DIR
/run/user/1000
syntezoid@workcomputer ~ 21:32 ❯ wayvnc
ERROR: Failed to initialise wayland
```

I delete WLR's lines from .pam_environment, from work computer, and logout and ssh again, but that didn't change anything.

**Reply by Flakebi** [reply]
I think you need to run wayvnc from inside sway (`WAYLAND_DISPLAY` and `SWAYSOCK` need to be set).
Usually, I run `env WLR_BACKENDS=headless WLR_LIBINPUT_NO_DEVICES=1 sway -c ~/.config/sway/vnc.config` where `vnc.config` is my sway config plus an `exec wayvnc 127.0.0.1 5900` at the end.

**Reply by phdpe** [reply]
> Usually, I run `env WLR_BACKENDS=headless WLR_LIBINPUT_NO_DEVICES=1 sway -c ~/.config/sway/vnc.config` where `vnc.config` is my sway config plus an `exec wayvnc 127.0.0.1 5900` at the end.

I was looking for a way to start wayvnc automatically, and ran into your post.  This is brilliant. Thank you.

**Reply by W4JEW** [reply]
@phdpe what was the magic in @Flakebi post that was the key to getting wayvnc to start automatically? I'm working on a Raspberry Pi running Raspberry Pi OS Bookworm. The default installation starts wayvnc as a userland process and wayvnc dies when you choose logout, so it's currently not possible to login via VNC as a different user.

I too am looking to find a way to start wayvnc automatically as a service such that anyone can login.

I was initially getting an error about XDG_RUNTIME_DIR not being set when trying to run:

```
/usr/bin/wayvnc --render-cursor --keyboard=us
```

So I set XDG_RUNTIME_DIR to /run/usr/1000 then ran wayvnc, and received the following:

```
pi@presto:~ $ /usr/bin/wayvnc --render-cursor --keyboard=us
ERROR: ../src/main.c: 402: WAYLAND_DISPLAY is not set in the environment
ERROR: ../src/main.c: 1495: Failed to initialise wayland
```

I checked to see if sway is installed - I see numerous packages that start with sway - but not sure that's the path I need to take.

You'll have to forgive me - I'm a bit of a n00b when it comes to Wayland and wayvnc.

Any advice you (or anyone else) can provide, would be greatly appreciated!

Cheers!

**Reply by any1** [reply]
@W4JEW We're working on things for the pi: https://github.com/RPi-Distro/raspi-config/pull/226

**Reply by W4JEW** [reply]
> @W4JEW We're working on things for the pi: https://github.com/RPi-Distro/raspi-config/pull/226

Wonderful! Following
